ui/src/commonMain/kotlin/com/materiiapps/gloom/ui/viewmodels/repo/tab/RepoDetailsViewModel.kt

```diff
@@ -42,4 +44,39 @@ class RepoDetailsViewModel(
         }
     }
 
+    fun toggleStar() {
+        fun updateStarDetails(starred: Boolean) {
+            details = details!!.copy(
+                viewerHasStarred = starred,
+                stargazerCount = details!!.stargazerCount + if (starred) 1 else -1
+            )
+        }
+
+        val details = details!!
+        val hasStarred = details.viewerHasStarred
+
+        // Optimistic update
+        updateStarDetails(starred = !hasStarred)
+
+        coroutineScope.launch {
+            isStarLoading = true
+
+            if (hasStarred) {
+                gql.unstarRepo(details.id)
+            } else {
+                gql.starRepo(details.id)
+            }.fold(
+                onSuccess = {
+                    isStarLoading = false
+                },
+                onError = {
+                    // Revert optimistic update
+                    updateStarDetails(starred = hasStarred)
+
+                    isStarLoading = false
+                    hasError = true
```

Comment:
hasError is only for if the details couldn't load, not for any generic error (I'll rename the variable later)